Fix multiline semantic highlighting for 'record'

hopehadfield opened the pull request:
Fixes redhat-developer/vscode-java#1444

Before:
![Screenshot from 2023-08-03 15-47-17](https://github.com/eclipse-jdtls/eclipse.jdt.ls/assets/115827695/f3e70dbd-bb21-4565-8ac4-296ccfba182c)

After:
![Screenshot from 2023-08-03 16-13-57](https://github.com/eclipse-jdtls/eclipse.jdt.ls/assets/115827695/b5f9b2dd-9c74-4396-b527-c23a64e685c3)

hopehadfield commented:
I believe this fixes redhat-developer/vscode-java#1442 as well, though I couldn't reproduce the automatic line-splitting for inner records mentioned, even with version 0.61.0.  The highlight colours are now correct.

rgrunber commented:
It looks very reasonable. I'll just have a look at some other way to properly apply the same colouring as `class` gets. It could be that it's only possible from the client side.

0dinD commented:
When calling `addToken`, you need to make sure that the tokens are added in the correct order (the order of appearance in the document). That's because the tokens are encoded using line/column deltas, and negative deltas are not supported since the Language Server Protocol specifies that they are unsigned integers.

Currently, you first add the `record` keyword token, and then you return `true`, which means that the visitor will continue visiting the child nodes of the `RecordDeclaration`, some of which appear *before* the keyword token, which will cause problems when encoding the tokens. For example, this PR causes a regression where the `public` modifier keyword before the `record` keyword is missing from the semantic tokens. I believe the same thing will happen with semantic tokens in the Javadoc comment of the `RecordDeclaration`, which also appears before the `record` keyword.

In fact, you can see that the tokens have been incorrectly encoded, because there is a test failure for `SemanticTokensHandlerTest#testSemanticTokens_Constructors`. First, it complains that it expected the `private` modifier keyword before the `record` keyword. And even if you add an assertion for the `record` keyword, it will complain that `Token deltaColumn should not be negative` for the `private` token.

---

Here is the correct visitation order: https://github.com/eclipse-jdt/eclipse.jdt.core/blob/4c26c0761103220355168e77cd6e556274b131a7/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/RecordDeclaration.java#L305-L312

The `record` keyword appears between the modifiers and the name/identifier, so I would propose something like this:

```java
	@Override
	public boolean visit(RecordDeclaration node) {
		acceptNode(node.getJavadoc());
		acceptNodeList(node.modifiers());

		// Here goes the code that calls addToken for the "record" keyword

		acceptNode(node.getName());
		acceptNodeList(node.typeParameters());
		acceptNodeList(node.recordComponents());
		acceptNodeList(node.superInterfaceTypes());
		acceptNodeList(node.bodyDeclarations());

		return false;
	}
```

0dinD commented:
Also, what's the reason for adding semantic token modifiers such as `public`, `static` etc to the `record` token? The way I see it, those modifiers should only apply to the name/identifier token, i.e. "Tuple" in your screenshots. The record itself may be `public`, `static` etc, but the `record` keyword is just a keyword, I don't think there is any need for further semantic information.

---

Regarding this code comment: `Token type 'MODIFIER' is used (rather than 'RECORD') to provide the correct highlight colour.`

The `RECORD` token type would not be correct for any keyword in any context, it is meant for a name/identifier token (and identifiers are not keywords). It is analogous to the `class`, `enum`, `interface` etc. token types: https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#standard-token-types-and-modifiers

0dinD commented:
@rgrunber:

> I'll just have a look at some other way to properly apply the same colouring as `class` gets. It could be that it's only possible from the client side.

The `class` declaration keyword is defined as `storage.modifier.java` in the TextMate grammar. That is to say, it has the same TextMate token type as the modifier keywords, `public`, `static`, `final` etc. In my opinion, this is incorrect (yet another problem with the Java TextMate grammar). If you look at a language like TypeScript, the `class` declaration keyword is defined to have the `storage.type.class.ts` TextMate token type. For reference, here are the TextMate conventions: https://macromates.com/manual/en/language_grammars#naming_conventions

But that feels like a separate discussion. The fact is that currently, the `class`, `record` etc. keyword is treated like a modifier keyword in the TextMate grammar. And so if we want the semantic tokens to be consistent with the TextMate tokens, we will have to use the `modifier` token type, as is currently done in this PR. It is mapped to the `storage.modifier.java` TextMate token: https://github.com/redhat-developer/vscode-java/blob/8fa2464e16028b248fd89efc14622aeab251e890/package.json#L148

hopehadfield commented:
@0dinD Thanks for the input. I've fixed the order of the token addition, removed the modifiers for the `record` token, and updated the `SemanticTokensHandlerTest#testSemanticTokens_Constructors` test so that it passes now.

rgrunber commented:
I suspect the choice for `class` was guided by Eclipse :

![image](https://github.com/eclipse-jdtls/eclipse.jdt.ls/assets/1417342/ebd65b0a-95a0-4860-8dbc-ceeff77a5433)

Thanks for the reference to  `semanticTokenScopes`. That's what I was curious to know.

I just have one question about the approach. If the semantic tokens must be added in order of appearance int the document, why don't we just sort them at the very end, instead of making the reading process more convoluted ? I can't imagine we're saving that much time by avoiding the additional sort at the end.

0dinD commented:
@hopehadfield:

> I've fixed the order of the token addition, removed the modifiers for the record token, and updated the `SemanticTokensHandlerTest#testSemanticTokens_Constructors` test so that it passes now.

Great, this PR looks good to me now! On paper, I haven't had time to test it.

It looks like the CI tests have not run for this PR (or at least, I can't see them linked here on GitHub, all I see is CodeQL and ECA). Do you have any idea why that might be @rgrunber? Maybe because this PR was a draft PR at first, and then got converted into a "normal" PR?

rgrunber commented:
test this please.

rgrunber commented:
Test this please.

0dinD commented:
I have some things to add to this discussion: https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/2789#issuecomment-1673332624

After thinking about it some more, I realize that we could probably just change the implementation of `addToken` so that it performs Insertion Sort while building the list. This should have a very minimal performance impact (an order of magnitude less than sorting the entire list after building it), because most tokens will be inserted in the correct order, due to the ordered nature of the AST and `ASTVisitor` (which gives Insertion Sort near-linear performance).

Also, think we can probably get rid of most (if not all) of the manual child-visitation logic (the `acceptNode` / `acceptNodeList` calls), and instead rely on `StructuralPropertyDescriptor`s in the AST. I hadn't realized until now that they can give us a lot of information about the relation between a child and its parent, using `ASTNode#getLocationInParent()`. This means that many of my previous workarounds that involve manually visiting child nodes from the parent can probably be refactored.

---

All of which is to say that I think there are a bunch of improvements that could be made to `SemanticTokensVisitor` to simplify the code and reduce the future maintenance burden. I'll investigate this more when I have time, and probably send a PR if all goes well.

---

But that discussion is quite off-topic from this PR. I still think that the current implementation of this PR is good, given the current state of `SemanticTokensVisitor` (where we need to manage token insertion order "manually").

0dinD commented:
> It seems only tokens not part of the AST model are at risk of being visited out of order

No, the risk of out-of-order visitation is introduced in all cases (even for tokens derived from AST nodes) where we avoid delegating the child visitation logic (returning `false` from a visitor method), and instead visit the children manually (via `acceptNode` / `acceptNodeList`).

As a reminder, here's one of my previous comments:

> **Nowadays**, we _do_ have to worry about visitation order, because we sometimes choose not to delegate to `ASTVisitor` for the child visitation logic. This is because we sometimes need to pass additional information from parent nodes down into the visitation logic for child nodes (due to bugs and limitations in the JDT Core API). I mean, an alternative solution would be for the child nodes to query information from their parent, but this quickly leads to more code complexity, because one parent has many child nodes (and even more grandchild nodes, and so-on).

Here's an example where there is a risk of out-of-order visitation even though we are only visiting AST nodes: https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/6fe54256bb644526fa58e8d90ba3a2ded92e7fbb/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/SemanticTokensVisitor.java#L316-L323

Now, I'm not sure if I explained what I meant very well in the above paragraph that I quoted. And I'm not sure if I agree with what I said anymore either way, now that I became aware of `StructuralPropertyDescriptor` and the information they can give us in terms of querying parent information from the child nodes. I think there might be ways in which the `SemanticTokensVisitor` code can be simplified, in regards to child visitation logic. It would be good if we could eliminate the risk of out-of-order visitation, by always delegating the child visitation logic. I'll investigate this in the future, when I have time.
